ct/l1: introduce simple metastore

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR lays the foundations for a replicated-STM-backed L1 metastore:
- initial data structures for the STM
- abstraction for updating the state safely
- an interface for the metastore
- an implementation of the metastore using the abstractions introduced here, but without replication or persistence

My goal in checking this in as is is to unblock other work that can make use of the metastore abstraction as is. See the individual commit messages for more details.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
ct: adjust visibility of a couple packages

Will be useful in upcoming changes:
- logger
- types

andrwng committed:
ct/l1: add initial STM state

Adds an initial cut of state for a L1 metastore. The introduced state
tracks a list of extents (references to L1 objects with some metadata to
allow for coarse-grained indexing) per Kafka partition, as well as the
objects themselves.

The mental model of metadata tracked per partition is pretty analogous
to the remote segments stored in cloud_storage::partition_manifest, with
the main differences:
- there are only Kafka offsets
- there are no terms stored in the L1 metadata
- rather than object names being derived from metadata, object names
  represented by an unique object_id that may be referred to by multiple
  extents

This commit only includes the data structures. A subsequent commit will
introduce abstractions to safely modify the state.

NOTE: the code here looks quite similar to datalake/coordinator/state.h.
Unlike the datalake coordinator, rather than allowing exactly one topic
revision to exist, here we allow for multiple topic reivisions to
coexist, with the expectation that one will eventually be removed.
Keeping a single topic revision active was more critical for datalake,
since each active topic needed to map to exactly one Iceberg table. No
such restriction exists here, and I suspect that being able to track
multiple revisions of a topic will simplify topic lifecycle management.

andrwng committed:
ct/l1: add initial add_objects state_update

Adds an abstraction to represent safely (e.g. no gaps, no overlap) add a
group of L1 objects to the `state`. This abstraction checks that
partition offset ranges are valid before allowing the operation to be
applied.

In subsequent work, the update abstraction will be used as replicated
commands for a replicated STM. As such, they are serde-able.

andrwng committed:
ct/l1: add metastore interface

Adds an interface for the L1 metastore. This is a fairly barebones
interface, currently only supporting a handful of calls (add_objects and
several getters), but this should be enough to e.g. get started on the
reconciler.

andrwng committed:
ct/l1: add simple_metastore

Adds an implementation of the metastore interface that is backed by the
newly introduced `state`. This metastore is not replicated or persisted
at all, but the state_updates that will be used for the STM are used
here as well in, so there should be little to no behavioral differences
to the eventual replicated-STM-backed metastore. Since metastore only
depends on simple struct definitions, it is ideal for parallelizing work
that doesn't require end-to-end testing.

dotnwat commented:
🔥

dotnwat commented:
merge skew

```

src/v/model/topic_revision.cc:13:10: error: 'absl/strings/str_split.h' file not found with <angled> include; use "quotes" instead
--
  | 13 \| #include <absl/strings/str_split.h>
  | \|          ^~~~~~~~~~~~~~~~~~~~~~~~~~
  | \|          "absl/strings/str_split.h"


```

rockwotj commented on src/v/cloud_topics/BUILD:
```diff
@@ -40,6 +41,7 @@ redpanda_cc_library(
         "types.h",
     ],
     include_prefix = "cloud_topics",
+    visibility = [":__subpackages__"],
```
We should have a discussion about code structure at some point

andrwng replied:
Yeah. Are you thinking to explicitly define a common library for submodules to use? In this case I only really wanted to use `object_id`. I think there are a few things in `types` that are more specific to L0.

rockwotj replied:
Yeah the `types.h` thing doesn't scale. This is fine for now, but I'll start a slack thread on how we can structure going forward.

rockwotj commented on src/v/model/topic_revision.h:
```diff
@@ -0,0 +1,82 @@
+        return {.topic = tp.topic, .rev = rev};
+    }
+};
+std::ostream& operator<<(std::ostream&, const tpr&);
```
prefer fmt::formatter for new code

andrwng replied:
Done

rockwotj commented on src/v/model/topic_revision.cc:
```diff
@@ -0,0 +1,47 @@
+namespace model {
+
+tpr tpr::from(std::string_view s) {
+    std::vector<ss::sstring> ss = absl::StrSplit(s, "_");
```
topic names can have `_` in them. Consider using `@` or you'll have to manually get the last segment here.

andrwng replied:
Ah 🤦 need to separate on / first

andrwng replied:
Will add a test tomorrow

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+// Represents the state managed by the replicated state machine that serves L1
+// metadata.
+
+// An object containing data for a contiguous range of batches belonging to a
```
It's not an object but a description of the object?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+    kafka::offset base_offset;
+    kafka::offset last_offset;
+    model::timestamp base_timestamp;
+    model::timestamp last_timestamp;
```
Do you mean `max_timestamp`?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+    auto serde_fields() { return std::tie(extents); }
+
+    // The list of extents that comprise this partition.
+    std::deque<extent> extents;
```
why deque?

andrwng replied:
We'll need to push to the back and pop from the front so it seemed suitable (though I forgot about chunked_fifo). Though thinking about it again, I think we'll also need to erase and add to the middle of the container for leveling/compaction. Swapped it out for an absl::btree_set. Longer term, this will likely be swapped with a store.

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+      envelope<partition_state, serde::version<0>, serde::compat_version<0>> {
+    friend bool operator==(const partition_state&, const partition_state&)
+      = default;
+    auto serde_fields() { return std::tie(extents); }
```
missing offsets?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+      envelope<topic_state, serde::version<0>, serde::compat_version<0>> {
+    friend bool operator==(const topic_state&, const topic_state&) = default;
+    auto serde_fields() { return std::tie(pid_to_state); }
+    absl::btree_map<model::partition_id, partition_state> pid_to_state;
```
`chunked_hash_map`?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+  : public serde::
+      envelope<object_entry, serde::version<0>, serde::compat_version<0>> {
+    friend bool operator==(const object_entry&, const object_entry&) = default;
+    auto serde_fields() { return std::tie(total_data_size, removed_data_size); }
```
missing `footer_pos`?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+  : public serde::envelope<state, serde::version<0>, serde::compat_version<0>> {
+    friend bool operator==(const state&, const state&) = default;
+    auto serde_fields() { return std::tie(topic_to_state, objects); }
+    absl::btree_map<model::topic_revision, topic_state> topic_to_state;
+    absl::btree_map<object_id, object_entry> objects;
```
`chunked_hash_map`?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/metastore.h:
```diff
@@ -0,0 +1,84 @@
+// the underlying implementation, whether it's an in-memory store, replicated
+// store, or even an external service.
+
+class metastore {
```
Consider briefly describing the metastore, or linking to the design document.

andrwng replied:
Done, added a high level overview. LMK if there's anything else you think would be helpful to include

rockwotj replied:
It's perfect thank you

rockwotj commented on src/v/cloud_topics/l1/metastore.h:
```diff
@@ -0,0 +1,84 @@
+    };
+
+    // Returns offsets (e.g. start, next) for the given partition.
+    virtual ss::future<std::expected<offsets_response, errc>>
```
`std::expected` :fire:

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+    object_id oid;
+};
+struct extent_lt_base_offset {
+    bool operator()(const extent& lhs, const extent& rhs) const {
```
nit: consider `std::strong_ordering operator<=>(const extent&) const;` instead?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state_update.h:
```diff
@@ -0,0 +1,91 @@
+        kafka::offset base_offset;
+        kafka::offset last_offset;
+        model::timestamp base_timestamp;
+        model::timestamp last_timestamp;
```
max_timestamp

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/state_update.h:
```diff
@@ -0,0 +1,91 @@
+    };
+
+    friend bool operator==(const new_object&, const new_object&) = default;
+    auto serde_fields() { return std::tie(oid, extent_metas); }
```
footer_pos

andrwng replied:
Done

rockwotj reviewed (commented):
Here's a quick pass. I haven't looked at the state logic or tests in depth yet.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a03762a9-b636-4781-b5ab-24b0f5c0f420-->
<!--TEST RESULTS-->
<!--55417fe7-5100-42e9-a5ee-b2224c69283f-->
<!--TEST RESULTS-->
<!--fd7cb0b6-412c-4ee5-ad14-48d2e7825ffe-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67945">build#67945</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------------|:------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest                  | test_connection_flicker                   | {"cloud_storage_type": 1}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67945#0197aa4b-46de-4bc4-b9e7-544acc8bb845 | FLAKY         | 20/21    | upstream reliability is '99.672131147541'. current run reliability is '95.23809523809523'. drift is 4.43404 and the allowed drift is set to 50. The test should PASS   |
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67945#0197aa50-9c3a-4151-853b-16768828f145 | FLAKY         | 15/21    | upstream reliability is '68.82217090069284'. current run reliability is '71.42857142857143'. drift is -2.6064 and the allowed drift is set to 50. The test should PASS |
| LeadershipPinningTest         | test_leadership_pinning_sanctions         |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67945#0197aa4b-46dc-492c-8d2a-1d6fef90548d | FLAKY         | 20/21    | upstream reliability is '99.6742671009772'. current run reliability is '95.23809523809523'. drift is 4.43617 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67945#0197aa50-9c3d-4c7b-ae04-be2928cc2d1a | FLAKY         | 20/21    | upstream reliability is '99.64539007092199'. current run reliability is '95.23809523809523'. drift is 4.40729 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68107">build#68107</a></summary>


| test_class                       | test_method                                   | test_arguments                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------|:----------------------------------------------|:---------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowIndexingLocalRetentionTest | test_local_time_based_retention_is_overridden | {"cloud_storage_type": 2, "local_retention_ms": 3600000} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68107#0197b332-156f-49cb-ae14-f04e6b7beff0 | FLAKY         | 20/21    | upstream reliability is '92.42819843342036'. current run reliability is '95.23809523809523'. drift is -2.8099 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68125">build#68125</a></summary>


| test_class                | test_method                             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:----------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DisablingPartitionsTest   | test_disable                            |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68125#0197b3e2-93fc-419f-8040-163d3afa603d | FLAKY         | 17/21    | upstream reliability is '91.65487977369166'. current run reliability is '80.95238095238095'. drift is 10.7025 and the allowed drift is set to 50. The test should PASS |
| RecreateTopicMetadataTest | test_recreated_topic_metadata_are_valid | {"replication_factor": 3} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68125#0197b3e2-93fc-419f-8040-163d3afa603d | FLAKY         | 19/21    | upstream reliability is '98.95615866388309'. current run reliability is '90.47619047619048'. drift is 8.47997 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

Lazin commented on src/v/model/topic_revision.h:
```diff
@@ -0,0 +1,82 @@
+    auto serde_fields() { return std::tie(topic, rev); }
+
+    model::topic topic;
+    revision_id rev;
```
should it be initial_revision_id? the revision_id changes with every partition move

andrwng replied:
It's true that partitions also have revisions, but the revision associated with a given topic never changes.

My goal in using revision_id is that it uniquely identifies the topic within the cluster. Given how initial_revision_id is used I worry that using it instead would not make it clear that that property holds.

rockwotj replied:
Why not use `topic_id`? I feel like there is some functionality that will appreciate having this be able to be manually assigned (where the revision_id is a property of when it's written the topic id is just an arbitrary value we can set).

andrwng replied:
>Given how initial_revision_id is used I worry that using it instead would not make it clear that that property holds.

Actually thinking about this more, the only time this can happen is during recovery, and if that's the case, I think it'd be fine to have a restored topic share the same key as the original -- presumably the metadata served by the restored topic is meant to be exactly the same as the original.

In any case, I like Tyler's suggestion of using topic_id here. Opened a Slack thread about it to see if there are any surprises that come about before making the change

andrwng replied:
Done

Lazin commented on src/v/cloud_topics/l1/BUILD:
could the name be `level_one`?

andrwng replied:
I'm happy to make the change, but curious is there any a particular reason? Should I change anything else to `level_one`?

dotnwat replied:
i think that perhaps this naming originates from Tyler's PR (probably because `l` and `1` are too similar looking)

rockwotj replied:
Yeah let's have a discussion about the cloud topics subsystem layout in slack or it can probably wait until our Wednesday meeting. I'm fine with moving the stuff in my PR to a `level_one` directory. +1 to `l` and `1` looking to similar in a lot of fonts.

andrwng replied:
Moved things over to `level_one`

Lazin commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+// A description of a portion of an object that points at data for a contiguous
+// range of batches belonging to a single partition. The object itself may
+// contain ranges of data from many partitions.
+struct extent
```
The struct is similar to `extent_meta` in `cloud_topics/extent_meta.h`. The `extent_meta` is supposed to be used to communicate with the data layer. It can't be serialized. But this struct can be serialized and it's supposed to be persisted. Maybe the name can reflect this to avoid confusion?

andrwng replied:
Do you think `extent_state` would be clearer?

Or are you suggesting I use something different from "extent" altogether -- like NTP chunk or slice or somesuch

dotnwat replied:
>Do you think extent_state would be clearer?

>Or are you suggesting I use something different from "extent" altogether -- like NTP chunk or slice or somesuch

i think `extent` is fine--it's an accurate name. but maybe we can converge on an answer by first deciding why `extent_meta` lacks timestamps if it can take on timestamps then we have equivalence (and just need to deal with serialization). if it it can't take on timestamps, then both names may benefit from adjustment.

Lazin commented on src/v/cloud_topics/l1/state_update.cc:
```diff
@@ -0,0 +1,106 @@
+    return update;
+}
+
+std::expected<std::monostate, stm_update_error>
```
Why `std::expected` is used here? So far we used `result` and `checked` in cloud topics and this adds another flavor of error handling to the mix.

andrwng replied:
As we continue to move towards newer versions of C++, it seems worthwhile to pick up the new standards as they introduce drop-in replacements for thirdparty things that we have. std::expected seems strictly better (e.g. it has monadic functions).

dotnwat replied:
+1 to std::expected. seems like it has everything we need.

rockwotj replied:
I'd be in favor of dropping `result` and `checked` in favor of `std::expected`. TBH I never think of when to use `result` vs `checked`.

Lazin commented on src/v/cloud_topics/l1/state_update.h:
```diff
@@ -0,0 +1,103 @@
+    size_t footer_pos;
+    chunked_hash_map<
+      model::topic_revision,
+      chunked_hash_map<model::partition_id, metadata>>
```
Do I understand correctly that we will not be able to put data from different topics into the same L1 object?

andrwng replied:
That's not correct -- note this is a two-level map, with topic and partition.

rockwotj replied:
Maybe I should update my data structure in the L1 object footer. I'm using an `absl::btree_multimap<model::ntp, footer_data>`.

Lazin commented on src/v/cloud_topics/l1/state_update.h:
```diff
@@ -0,0 +1,103 @@
+      = default;
+    auto serde_fields() { return std::tie(new_objects); }
+
+    static constexpr auto key{update_key::add_objects};
```
that should probably be another key

andrwng replied:
I'm not sure what you mean. Can you elaborate? Do you mean `key` should be different, or `add_objects`?

dotnwat commented on src/v/model/topic_revision.h:
```diff
@@ -0,0 +1,82 @@
+    }
+};
+
+struct tpr
```
looks fine to me. should we ping replication team as presumably they'll also be adopting this type name? separately, i'm wondering if we can just add the revision to `model:topic_partition`?

andrwng replied:
Removed this in favor of using topic_id, but also introduced topic_id_partition

dotnwat commented on src/v/model/topic_revision.h:
```diff
@@ -0,0 +1,82 @@
+
+    topic_revision topic_revision() const {
+        return {.topic = tp.topic, .rev = rev};
+    }
```
this won't return the topic revision, right? it'll return the partition's revision as-if it is the topic revision?

andrwng replied:
Ah I'll clarify this in comments and maybe rename rev to `topic_rev` (unless we go with something topic_id-based) -- the revision here is the topic revision, not the partition revision.

andrwng replied:
No longer relevant with the move to topic_id

dotnwat commented on src/v/model/topic_revision.cc:
```diff
@@ -0,0 +1,35 @@
+        throw std::runtime_error(fmt::format("Invalid TPR: {}", s));
+    }
+    int pid = std::stoi(pr[0]);
+    int rev = std::stoi(pr[1]);
```
i think the revision is an int64, so you probably want std::stoll here?

rockwotj replied:
I'm biased, but you're already using `absl` I find `absl::SimpleAtoi<type>` to be more explicit and clear.

andrwng replied:
Done

dotnwat commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+
+// State tracked per Kafka partition. The extents added to this state must have
+// no overlaps and no gaps in order to ensure there is no data loss.
+struct partition_state
```
i'm slightly confused by these types being serializable. i was expecting to see only types used for in-memory state representation in this PR. are you thinking that the state we represent in memory is very similar to that which will be eventually persisted as raft commands or into the backing store?

andrwng replied:
>are you thinking that the state we represent in memory is very similar to that which will be eventually persisted as raft commands

Yeah, minus any optimizations (e.g. using a cstore) I expect this state will roughly be used as the Raft snapshot when implementing a replicated STM

dotnwat commented on src/v/cloud_topics/l1/state_update.h:
```diff
@@ -0,0 +1,103 @@
+
+using stm_update_error = named_type<ss::sstring, struct update_error_tag>;
+
+struct new_object
```
so this is something that would go into a raft command (like L1 registration command?)

andrwng replied:
Right, exactly. I intend on having the serialized `add_objects_update` be the STM's raft payload

dotnwat commented on src/v/cloud_topics/l1/metastore.h:
```diff
@@ -0,0 +1,104 @@
+    // Adds the given objects to the metastore. If any are invalid, e.g.
+    // because they break an invariant of a partition's offset ranges, all
+    // objects are rejected.
+    virtual ss::future<std::expected<std::monostate, errc>>
```
i'm pretty sure with `std::expected` you can use the void specialization `std::expected<void, errc>`.

andrwng replied:
Yeah, but it's actually kinda ugly:

```
co_return std::expected<void, metastore::errc>{};
```

vs just co_return `std::monostate{}`. I can make the change if you want, or maybe I'm missing a cleaner way to co_return...

dotnwat reviewed (commented):
yeh i think this all look pretty good. i'm slightly confused by many of the types in this PR being serde-enabled (as opposed to having an in-memory representation, and then a set of specific types serialized to raft log), but maybe I'm missing a bigger part of the story.

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,130 @@
+    auto serde_fields() { return std::tie(extents, start_offset, next_offset); }
+
+    // The list of extents that comprise this partition.
+    absl::btree_set<extent> extents;
```
do you want this to be ordered? Maybe a comment?

andrwng replied:
Yeah, sortedness will be useful to find an element with a specific offset, and to remove elements from the start of the set (applying retention).

rockwotj reviewed (commented):
A few more comments, will finish a full review tomorrow

andrwng committed:
model: add topic_id_partition

Adds a new topic_id_partition, similar to the existing topic_partition,
but uniquely identifies the revision of the topic as well.

I intend on using this to keep track of L1 metadata per partition, in
which case it's critical to distinguish between revisions of a topic
name.

andrwng committed:
ct/l1: move l1 -> level_one

Review feedback: it's easier to read.

andrwng commented:
Diff since last round of reviews: https://github.com/redpanda-data/redpanda/compare/8d2f847f07c1142f2e923d06b38778008e2f2e41..d952ad9f9ef96bb98672ca7a43ef79864b9981b3

andrwng commented on src/v/cloud_topics/level_one/simple_metastore.cc:
```diff
@@ -0,0 +1,105 @@
+new_object make_new_object(const metastore::object_metadata& o) {
+    new_object new_o{
+      .oid = object_id{o.oid()},
+    };
```
Done

andrwng commented on src/v/cloud_topics/level_one/simple_metastore.cc:
```diff
@@ -0,0 +1,105 @@
+    // TODO: use binary search, given the offsets in the extents are guaranteed
+    // to be monotonically increasing.
+    auto& prt = prt_ref->get();
+    for (const auto& obj : prt.extents) {
```
Done

rockwotj commented on src/v/model/fundamental.h:
```diff
@@ -527,6 +528,54 @@ using topic_id = named_type<uuid_t, struct topic_id_type>;
+
+    friend std::ostream& operator<<(std::ostream&, const topic_id_partition&);
+
+    friend void read_nested(
+      iobuf_parser& in, topic_id_partition& tp, const size_t bytes_left_limit) {
+        using serde::read_nested;
+
+        read_nested(in, tp.topic_id, bytes_left_limit);
+        read_nested(in, tp.partition, bytes_left_limit);
+    }
+
+    friend void write(iobuf& out, topic_id_partition tp) {
+        using serde::write;
+
+        write(out, tp.topic_id);
+        write(out, tp.partition);
+    }
```
Why not `serde_fields`?

andrwng replied:
Somewhat this was copy-pasta from topic_partition, but also it feels like such an unlikely thing to be changed that it didn't seem worth the (admittedly small) extra overhead of an envelope

rockwotj commented on src/v/model/fundamental.h:
```diff
@@ -527,6 +528,54 @@ using topic_id = named_type<uuid_t, struct topic_id_type>;
+    bool operator<(const topic_id_partition& other) const {
+        return topic_id < other.topic_id
+               || (topic_id == other.topic_id && partition < other.partition);
+    }
```
`==` can be defaulted, `!=` and `<` are not needed because they can be auto derived from `!=` and `<=>`

andrwng replied:
Done

rockwotj commented on src/v/model/tests/topic_id_partition_test.cc:
```diff
@@ -0,0 +1,33 @@
+    ASSERT_EQ(12345, tidp.partition());
+}
+
+TEST(TopicIdPartitionTest, TestNoPartition) {
```
empty partition test?

andrwng replied:
Not sure if you mean this should have a different name, or a new test case. Assuming the latter and added one where there's a slash but no partition

rockwotj commented on src/v/cloud_topics/l1/state.h:
```diff
@@ -0,0 +1,135 @@
+
+    kafka::offset base_offset;
+    kafka::offset last_offset;
+    model::timestamp min_timestamp;
```
why do we need a min timestamp? I guess we can remove it later, but I think based on kafka timequery semantics we only need the max.

andrwng replied:
DOne

rockwotj commented on src/v/cloud_topics/l1/state_update.h:
```diff
@@ -0,0 +1,103 @@
+
+        kafka::offset base_offset;
+        kafka::offset last_offset;
+        model::timestamp min_timestamp;
```
Same question about `min_timestamp` I don't think it's useful for timequeries.

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/l1/metastore.h:
```diff
@@ -0,0 +1,103 @@
+// store, performing atomic updates to state and answering simple queries for
+// callers. The metastore itself isn't meant to perform complex partition
+// storage management tasks (e.g. garbage collection or compaction), but rather
+// it exposes primitives to enable robustly performing such tasks.
```
Separate question we should take into slack: if GC, compaction and leveling are all distinct fibers, how do we ensure that they don't stomp on each other?

andrwng replied:
I don't have a strong opinion yet, but my rough thoughts atm are that the metastore relies entirely on optimistic concurrency, and scheduling of these operations is done in a way that plays well with that (e.g. scheduling one compaction or leveling operation per NTP at a time)

rockwotj commented on src/v/cloud_topics/l1/tests/simple_metastore_test.cc:
```diff
@@ -0,0 +1,338 @@
+                   .add(tid_a, 1_o, 10_o, 1000_t, 2000_t, 0, 99)
+                   .build();
+    auto add_res = m.add_objects(om_list_t::single(std::move(ometa))).get();
+    ASSERT_FALSE(add_res.has_value()) << int(add_res.error());
```
If we're going to use `std::expected` more it might be worth some matches for it: https://google.github.io/googletest/reference/matchers.html

Would allow us to write this here:

```c++
EXPECT_THAT(add_res, HasError(invalid_request));
EXPECT_THAT(add_res, HasValue(std::monostate{}));
```

Don't do this now, but something to consider for another PR

rockwotj commented on src/v/cloud_topics/l1/tests/simple_metastore_test.cc:
```diff
@@ -0,0 +1,338 @@
+
+model::timestamp operator""_t(unsigned long long t) {
+    return model::timestamp{static_cast<int64_t>(t)};
+}
```
🥇 maybe we should formalize this for tests

andrwng replied:
Yea, I like these a lot. Will consider this for a follow up

rockwotj commented on src/v/cloud_topics/l1/tests/simple_metastore_test.cc:
```diff
@@ -0,0 +1,338 @@
+    simple_metastore m;
+    om_list_t os;
+    os.emplace_back(om_builder(oid1, 100)
+                      .add(tid_a, 0_o, 10_o, 1000_t, 2000_t, 0, 99)
```
consideration for future/other work: having simple named structs here helps make the tests more readable (don't spend time changing now)

andrwng replied:
Yea i started out with this and ended up making these builders because i thought they'd look nicer but ended up not quite paying off. Will continue with named structs

rockwotj commented on src/v/cloud_topics/l1/metastore.h:
```diff
@@ -0,0 +1,103 @@
+
+    // Finds the first object of a given partition with data greater than or
+    // equal to the given timestamp. If no such timestamp exists, returns
+    // `invalid_request`.
```
I think a better name than `invalid_request` could be used for missing or out of range queries.

andrwng replied:
Done

rockwotj reviewed (commented):
LGTM, just a few nits and questions

rockwotj commented on src/v/cloud_topics/level_one/metastore.h:
```diff
@@ -0,0 +1,103 @@
+    get_offsets(const model::topic_id_partition&) = 0;
+
+    // Adds the given objects to the metastore. If any are invalid, e.g.
+    // because they break an invariant of a partition's offset ranges, all
```
I am not quite sure we are able to do leveling or compaction with this interface, but we can add APIs to remove or replace offset ranges in a followup PR

andrwng replied:
That's correct, the API added is just for appending to each NTP. Replacing ranges will come in a follow up

rockwotj commented on src/v/model/fundamental.h:
```diff
@@ -527,6 +528,42 @@ using topic_id = named_type<uuid_t, struct topic_id_type>;
+    bool operator==(const topic_id_partition& other) const = default;
+    auto operator<=>(const topic_id_partition& other) const noexcept = default;
+
+    friend std::ostream& operator<<(std::ostream&, const topic_id_partition&);
```
nit: use fmt::formatter

rockwotj replied:
This can go in a followup

rockwotj reviewed (approved):
Nice, just a couple nits. LGTM
